### PR TITLE
Adaptive Routing and Switching

### PR DESCRIPTION
**Pull request by JaiOCP**
This PR introduces a new concept where forwarding is done based parametric quality computation of the forwarding paths.

Following PRs have been closed because of enum churn but have all the audit trail of review comments. Please refer to them for history.
https://github.com/opencomputeproject/SAI/pull/1658
https://github.com/opencomputeproject/SAI/pull/1595

### CODE REVIEW

**Comment by JaiOCP** [comment]
@ashutosh-agrawal @eddyk-nvidia @marian-pritsak @rck-innovium @lguohan @rlhui 

Please review. This is final update based on all the feedbacks.

**Review by tushar-ty — CHANGES_REQUESTED** [review]

**Comment by eddyk-nvidia** [comment]
> @ashutosh-agrawal @eddyk-nvidia @marian-pritsak @rck-innovium @lguohan @rlhui
> 
> Please review. This is final update based on all the feedbacks.

@marian-pritsak and me are OK with the final release. Thanks

**Comment by ashutosh-agrawal** [comment]
> @ashutosh-agrawal @eddyk-nvidia @marian-pritsak @rck-innovium @lguohan @rlhui
> 
> Please review. This is final update based on all the feedbacks.

I'm okay with the proposal too.

File: doc/ARS/Adaptive-Routing-and-Switching.md [hunk 1]
```diff
@@ -0,0 +1,853 @@
...
+4. Minimize traffic drop when link flaps by HW based link down detection and path reassignment
+
+
+### 3.1 Terms and Acronyms
+| Acronym   | Term      |
+|:----------|:----------|
+| Micro Flow    | A single traffic flow typically identified by 5 tuple    |
+| Macro Flow    | collection of micro flows that hash to same value   |
+| Idle Time | Time when there are no packets observed within a macro flow |
+| Flowlet    | contagious subsequence of packets within a macro flow separated by an idle time interval    |
```
[the remark this answers is not in the document]
Comment (JaiOCP) [reply doc/ARS/Adaptive-Routing-and-Switching.md]: https://github.com/opencomputeproject/SAI/pull/1681/commits/c30f206e2feb7e2c75853bbdfbbb1626657a4ec7

File: doc/ARS/Adaptive-Routing-and-Switching.md [hunk 2]
```diff
@@ -0,0 +1,853 @@
...
+8. Link Failure or Admin Down
+*Besides being able to choose best quality path from the NHG, packet should also be able to select best link available amongst LAG memembers*
+
+> *Additional algorithms and quality metrics can be added laters*
+
+### 3.3 ARS path selections modes
+
+1. Flow-let based 
+*best quality path till the flow-let hits an idle time*
+*random path with out considering path quality*
```
[the remark this answers is not in the document]
Comment (JaiOCP) [reply doc/ARS/Adaptive-Routing-and-Switching.md]: https://github.com/opencomputeproject/SAI/pull/1681/commits/c30f206e2feb7e2c75853bbdfbbb1626657a4ec7

File: doc/ARS/Adaptive-Routing-and-Switching.md [hunk 3]
```diff
@@ -0,0 +1,853 @@
...
+3. Path reassignment takes flow (in)activity into account
+*Best path is selected when the flow starts. Path is locked while flow is active.*
+*Flow becomes eligible for path reassignment after the idle time.*
+*Idle time should be configured typically as 1/2 of RTT*
+
+4. Algorithm for load computation
+*EWMA is used for computing the historical and future load*
+
+5. Quantization
+*Computed value need to be quantized into buckets or bands for easier intepretation*
```
[the remark this answers is not in the document]
Comment (JaiOCP) [reply doc/ARS/Adaptive-Routing-and-Switching.md]: https://github.com/opencomputeproject/SAI/pull/1681/commits/c30f206e2feb7e2c75853bbdfbbb1626657a4ec7

**Comment by JaiOCP** [comment]
Thank you @marian-pritsak @eddyk-nvidia @ashutosh-agrawal @rck-innovium for providing your inputs. 
Can you please approve the PR for merge.

**Review by eddyk-nvidia — APPROVED** [review]

**Comment by JaiOCP** [comment]
@ashutosh-agrawal @rck-innovium Gentle reminder to approve this PR

**Review by ashutosh-agrawal — APPROVED** [review]

**Comment by JaiOCP** [comment]
@rck-innovium Please approve

**Pull request merged**